src/Take.Elephant.Redis/RedisSetMap.cs

```diff
@@ -1,70 +1,136 @@
             : base(mapName, connectionMultiplexer, db, readFlags, writeFlags)
         {
-            if (serializer == null) throw new ArgumentNullException(nameof(serializer));
-            _serializer = serializer;
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
             _useScanOnEnumeration = useScanOnEnumeration;
+            _supportEmptySets = supportEmptySets;
+            _emptyIndicatorExpiration = emptyIndicatorExpiration ?? TimeSpan.FromMinutes(15);
         }
 
+        public bool SupportsEmptySets => _supportEmptySets;
+
+        // Some methods below use tasks instead of async-await to 
+        // avoid redis transaction deadlock issues.
+        // See https://stackoverflow.com/questions/25976231/stackexchange-redis-transaction-methods-freezes
+        // TODO: perhaps we should slap async in everything here (see "Async guidance performance optimization" in our internal wiki)
+        // and leave not awaiting these methods as a responsibility for the caller
+        // since they are the ones who know whether or not the provided IDatabase
+        // is an instance of ITransaction
+
         public override async Task<bool> TryAddAsync(TKey key,
             ISet<TItem> value,
             bool overwrite = false,
             CancellationToken cancellationToken = default)
         {
-            if (key == null) throw new ArgumentNullException(nameof(key));
-            if (value == null) throw new ArgumentNullException(nameof(value));
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
 
-            var internalSet = value as InternalSet;
-            if (internalSet != null) return internalSet.Key.Equals(key) && overwrite;
+            if (value is InternalSet internalSet)
+            {
+                return internalSet.Key.Equals(key) && overwrite;
+            }
 
-            var database = GetDatabase() as IDatabase;
-            if (database == null) throw new NotSupportedException("The database instance type is not supported");
+            if (!(GetDatabase() is IDatabase database))
+            {
+                throw new NotSupportedException("The database instance type is not supported");
+            }
 
             var redisKey = GetRedisKey(key);
-            if (await database.KeyExistsAsync(redisKey, ReadFlags) && !overwrite) return false;
+            if (await database.KeyExistsAsync(redisKey, ReadFlags).ConfigureAwait(false) && !overwrite)
+            {
+                return false;
+            }
 
             var transaction = database.CreateTransaction();
+            var emptySetRedisKey = RedisSet<TItem>.GetEmptySetIndicatorForKey(redisKey);
+
+            if (value == null && _supportEmptySets)
+            {
+                var commandTask = transaction.StringSetAsync(emptySetRedisKey, true.ToString(), _emptyIndicatorExpiration);
```

Comment:
There is not need to use a transaction here, you can call just StringSetAsync, can't you?